### ISSUE TITLE
Native support for code snippets, code blocks, inline code using back ticks e.g. `code`

### ISSUE THREAD

**Issue by ElijahLynn** [tootsuite/mastodon#13544]
### Pitch

Add native support for code snippets, code blocks, inline code using back ticks e.g. `code`. 

### Motivation

When posting a toot, I want my code snippets that I surround with backticks (industry standard) to be formatted differently than regular, surrounding text. 

![image](https://user-images.githubusercontent.com/1504756/80264495-351c4080-8649-11ea-9ee1-1aa3a797805f.png)

**Comment by ThibG** [comment]
![image](https://user-images.githubusercontent.com/384364/80289517-244df680-873f-11ea-84f3-4a2a35c0aaf1.png)

(See #10629)

**Comment by Sasha-Sorokin** [comment]
>tl;dr; Very hot topic, people didn't accept boundaries of italic, bold and code subset and demanded literally all the formatting available: from headers to lists. Ended up with nothing, nor expecting anything in the future, though [there's still a possibility](https://github.com/tootsuite/mastodon/pull/10629#issuecomment-501741304).

It's actually quite sad, I'd be so happy about this subset of Markdown, and think other formatting is very redundant for ‘toots’ format, but could have been used for Article's as [in Hometown fork](https://github.com/hometown-fork/hometown#reading-more-content-types): Article objects can be sent by Write.as, so you could have read them right without leaving Mastodon.

<details>
<summary>Click for bits about reinventing the wheel</summary>

---

Speaking further, people seemed to misinterpret Eugen and thought he is going to invent another syntax, which caused some frustration. And even if that is what was supposed to happen, as a Telegram user I used to this, we had a special markdown that consisted of `*bold*`, `_italic_` and `` `code` `` ([ref.](https://core.telegram.org./bots/api#markdownv2-style)). For ‘toots’ it actually seems like a not-so-bad idea, giving limit of 500 characters.

... Though later message box became a [WYSIWYG](https://en.wikipedia.org/wiki/WYSIWYG) + would accept only regular syntax of Markdown, and instead special entities were sent by client, so you need not worry about this anymore, except if you created a bot, but you could use HTML there. No \*sarcasm\* was replaced with **sarcasm** by accident anymore.

Discord on the other hand sent a feature in beta that would solve the same problem of unintended formatting, but without removing a syntax, it would just highlight raw syntax with **\*\*bold\*\*** and *\*italic\**, `` `code` `` and show regular emojis as images, not `:their_names:`.

---
</details>

Eh, I wish we given this thing another round of discussion and people weren't as rude and demanding as before.

**Comment by AverytheFurry** [comment]
Yes, please add.

**Comment by brendanjones** [comment]
Agree, would love to have this. Though I presume if this can be implemented then all markdown can be implemented, in which case refer to https://github.com/mastodon/mastodon/issues/18958.

**Comment by ioquatix** [comment]
If we can agree to a supported subset of features I'd be happy to have a go at implementing this, but I'd need to get an agreement from someone authoritative on the matter otherwise I feel like I might do all the work for nothing.

**Comment by solnic** [comment]
Huge 👍🏻 I'm also happy to help.

**Comment by fredrikaverpil** [comment]
> I'd need to get an agreement from someone authoritative on the matter otherwise I feel like I might do all the work for nothing

@Gargron care to give an indication of how plausible this is to be accepted and any guidelines?

**Comment by joemicmc** [comment]
I would love this as a feature!

**Comment by chenasraf** [comment]
Please add this!

**Comment by nia-e** [comment]
+1, willing to attempt an implementation if this is given the go-ahead

**Comment by JohnDowson** [comment]
I don't see why a slippery slope to supporting full markdown is an argument against code blocks. First of all, you can just stop at code blocks, and second of all full markdown support would be good actually?

**Comment by MightySpaceman** [comment]
We really need this especially mastodon, even more than other platforms needs this due to it's larger...well...techie ratio

**Comment by benjaoming** [comment]
It's important to support this issue without scope creeping the feature proposal: This issue is _not_ about introducing Markdown support and it is NOT about adding syntax highlighting. It's only about being able to display `monospace code snippets` in a more readable way.

Element for Matrix, Slack and many other chat clients support this syntax without supporting Markdown.

**Comment by benjaoming** [comment]
Incoming text renders both `code` and `pre`, so perhaps we are getting closer? See: https://github.com/mastodon/mastodon/pull/23913

**Comment by nfoert** [comment]
+1 for this, has it been discussed any further?

**Comment by brawaru** [comment]
Markdown (which I assume will include code blacks) is in ‘Exploring’ section on https://joinmastodon.org/roadmap.

**Comment by grand-lotus-iroh** [comment]
Is this still not a thing? If I type (3) tilde symbols nothing works...

🤯

**Comment by rochacbruno** [comment]
https://elk.zone can do it and it is very helpful

![Screenshot_20240326_195542](https://github.com/mastodon/mastodon/assets/458654/33cd825e-c8d6-4557-b532-655f089660ae)

**Comment by brendanjones** [comment]
That's awesome! Unfortunately doesn't render in normal Mastodon:

<img width="590" alt="Screenshot 2024-03-27 at 19 13 54" src="https://github.com/mastodon/mastodon/assets/16049594/13ddbdf1-aaa2-467c-a76c-e0698a6c7114">

I wonder how they're doing it, and if it can be imported easily into the Mastodon codebase?

**Comment by benjaoming** [comment]
@brendanjones a lot of clients have client-side support for parsing posts, enabling certain Markdown-ish features. This issue is talking about inline code and code blocks _only_ support in Mastodon's clients (?). A client feature would mean to both display such Markdown syntax the correct way, and also to possibly add UI guidance for using it. As mentioned in other comments, this is in the Mastodon roadmap as well (MAS-86).

The Mastodon spec has been expanded to allow `<pre>` and `<code>` here: https://docs.joinmastodon.org/spec/activitypub/#sanitization

**Comment by karladler** [comment]
can we have this pleeeeeeease?

**Comment by lloydjatkinson** [comment]
Just posted something assuming it would exist (because why wouldn't it) and was disappointed.

![Image](https://github.com/user-attachments/assets/bfebdf4a-3079-4cb2-9406-b0f5286825f4)

Edit: of course the misery guts maintainers mark this as spam just because I put a funny meme

**Comment by JohnDowson** [comment]
Man, someone must really hate rich text if this is still open years later. It'd be a very small feature to implement otherwise.

**Comment by andypiper** [comment]
note that comments are welcome if they contribute to the issue itself; "me too", observations on timescales, and related additions are not helpful. Please contribute meaningfully and thoughtfully to avoid wasting the time of the maintainers and developers. Thank you.

**Comment by ChristophHaag** [comment]
A secondary feature would be to disable stuff like tags inside code blocks.

I just made a post with tags `#ifndef`, `#define`, and `#endif`. I'm not sure how useful those tags will be.
